### PR TITLE
Implement `helm pull` for OCI registries

### PR DESCRIPTION
**Pull request by pmengelbert**
This PR is in line with the HIP for OCI Proposal here: https://github.com/helm/community/pull/136

* New `helm pull` does not depend on registry cache
* Pull OCI-stored chart with `helm pull oci://<registryURL> --version <tag>`
* Implement `helm dep update` for oci dependencies, in this format:
```
dependencies
  - name: ocichart
    version: 0.1.0
    repository: "oci://localhost:3000/myregistry"
# equivalent to "oci://localhost:3000/myregistry/ocichart:0.1.0
```
* New unit tests for `helm pull` and `helm dep update`
* Implement `OCIGetter` implementation of `Getter` interface

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

**What this PR does / why we need it**:

**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

### CODE REVIEW

File: pkg/repo/repotest/server.go
Comment (jdolitsky) [inline pkg/repo/repotest/server.go:46]: this import should be up w/ the others

File: pkg/repo/repotest/server.go
Comment (jdolitsky) [inline pkg/repo/repotest/server.go:420]: should import and use `github.com/phayes/freeport` (which is already in go mod)

**Review by jdolitsky — CHANGES_REQUESTED** [review]

File: pkg/downloader/manager.go
Comment (pmengelbert) [reply pkg/downloader/manager.go:722]: The linter run by `make test` didn't like the naked return

File: internal/resolver/resolver.go
Comment (pmengelbert) [reply internal/resolver/resolver.go:121]: `vs` is looped over later, on line 140.

File: internal/resolver/resolver.go
Comment (pmengelbert) [reply internal/resolver/resolver.go:111]: yep, line 140

**Comment by bacongobbler** [comment]
Hi @pmengelbert and @jdolitsky. Have you reached out to @phroggyy on this discussion topic? Based on [his last comment](https://github.com/helm/helm/pull/7613#issuecomment-654686203) it did appear that he was still interested in pursuing this effort. I'm curious why another competing PR is being proposed here instead of working together with @phroggyy.

**Comment by jdolitsky** [comment]
No ill intentions here - just trying to move things forward based on the new proposal. (Let it be known that I had approved those changes back in February)

I've reached out to Leo in Slack. and will wait for response. We're happy to drop this PR if that one is updated to reflect the proposal.

**Review by jdolitsky — APPROVED** [review]

File: cmd/helm/pull.go
Comment (jdolitsky) [reply cmd/helm/pull.go:74]: ^

File: pkg/action/pull.go
Comment (jdolitsky) [reply pkg/action/pull.go:46]: ^

**Review by jdolitsky — CHANGES_REQUESTED** [review]
Please see comment regarding the `OCI` boolean, should not be needed

**Comment by phoenix-bjoern** [comment]
If this feature is removed again from roadmap for the next release because of „new considerations“ I am going nuts. We are waiting for progress on the OCI support for almost a year and „new options“ simply show that people are tired of waiting. 
Please make your choice and move on. There are more things to solve like recursive dependencies but without some progress we get really stuck here. 
Thanks.

**Comment by phroggyy** [comment]
@bacongobbler to clear up on my end: I have no issues with this going forward. I did not see your link to the HIP in my PR until I read Josh's message today, and so even without checking the code, I'd assume this is a lot closer to done based on that spec. As I said to Josh in DM, if any effort should "go to waste" it should be the older, more faulty one, imho.

**Comment by phroggyy** [comment]
@phoenix-bjoern I'd consider this a massive step forwards! Sure, it's frustrating that it's been open for so long, but let's keep in mind that there was no spec for this back in #7613. Then a spec (HIP) was written, and now there's a compliant implementation – this to me seems like a massive step forwards and I would believe brings us a lot closer to getting this feature in!

**Comment by phoenix-bjoern** [comment]
@phroggyy thanks for comment and all your efforts. Any implementation is fine for us as long as it means we do not have to wait another 3 months till OCI dependencies are supported in a GA version.

**Comment by bacongobbler** [comment]
@phoenix-bjoern while I understand the frustration, I want to make things clear... I was asking to make sure that @phroggyy was okay with this PR. Since he's put in the most time into this, we want to make sure that credit to the original author is due. If he's okay with @pmengelbert moving this forward (as he's done so), then that's fine with me. It's important to us to make sure that the original author is given credit when they put in the work.

Even after this PR is merged, it is highly unlikely this feature will be marked as stable. If you read the proposed HIP 6 at https://github.com/helm/community/pull/136, we still need the OCI to approve the OCI Artifacts proposal until we can mark this as a stable feature.  Please refer to Section 7: "Experimental until clear messaging from OCI" for more information.

If I were to assume a timeline, I wouldn't be expecting this to be marked as a stable feature until Helm 4.

Hope that clears things up.

**Review by jdolitsky — APPROVED** [review]

**Comment by phoenix-bjoern** [comment]
@bacongobbler the OCI feature is marked as experimental, clear. The OCI dependency issue, which I've created in #6593, has celebrated already it's first birthday ;-)
We now have two(!) working solutions. If everyone favors this implementation, could you at least agree to add this issue to the milestone list for 3.4.0? @phroggyy 's implementation in #7613 was planed for 3.2, 3.3 and 3.4, where it has been removed lately. So I just kindly ask, also for other folks who voted for this feature already in my initial issue, for some progress in this OCI functionality in order to be usable without additional workarounds/commands. Thank you.

**Comment by bacongobbler** [comment]
We've been focused on the Helm 2 deprecation, which happens in less than a month's time from now. Any Helm 3 work has been put on the side-burner for the time being while we support those folks still migrating to Helm 3. Thanks for your patience.

**Comment by jdolitsky** [comment]
FYI - The HIP related to this change has been accepted. Please see https://github.com/helm/community/blob/master/hips/hip-0006.md

**Review by WyriHaximus — APPROVED** [review]
Tested this out locally and it worked fine, aside from a file deletion error after the first time installing a chart.

One question though. Will this support semver versions ranges such as `^1.2` in the future? So we don't have to update the version we want by hand

**Comment by jdolitsky** [comment]
@WyriHaximus can you expand on "file deletion error after the first time installing a chart" ? The support for version ranges might be supported in the future if we are able to query the registry for available versions (tag listing).

**Comment by WyriHaximus** [comment]
> @WyriHaximus can you expand on "file deletion error after the first time installing a chart" ? 

@jdolitsky the exact error is:
```
Error: unable to move current charts to tmp dir: link error: cannot rename charts to tmpcharts: rename charts tmpcharts: file exists
```

> The support for version ranges might be supported in the future if we are able to query the registry for available versions (tag listing).

Was expecting as such, don't know OCI well enough to know if this is build in. But from a user perspective this would be very welcome and put it at the same level of use as the old skool `index.yaml` based repositories.

**Comment by jdolitsky** [comment]
The "tmpcharts" piece seems unrelated: https://github.com/helm/helm/blob/af2d3021f491838c0a8a396d5c6fdaa172732bd2/pkg/downloader/manager.go#L247

Unsure if this is an existing bug or specific to this change

**Comment by jdolitsky** [comment]
Seems people have had this issue in the past @WyriHaximus https://github.com/helm/helm/issues?q=is%3Aissue+cannot+rename+charts+to+tmpcharts+is%3Aclosed

**Comment by WyriHaximus** [comment]
@jdolitsky Looks like I had some naming crossed inside the OCI repository. But name made sure that the chart name is be found under `oci://repository.host.name` as `oci://repository.host.name/helm-oci-test` and a tag as `oci://repository.host.name/helm-oci-test:0.5.0`.

Getting this error now:
```
Error: unable to move local charts to charts dir: link error: cannot rename tmpcharts/helm-oci-test to charts/helm-oci-test: rename tmpcharts/helm-oci-test charts/helm-oci-test: file exists
```

When running this command:
```
~/Projects/Helm/helm/bin/helm dependency update
```

Now this happens as long as the `charts/helm-oci-test` directory exists. When I remove `charts/helm-oci-test` but leave `charts` present the error becomes:

```
Error: unable to move current charts to tmp dir: link error: cannot rename charts to tmpcharts: rename charts tmpcharts: file exists
```

Not sure how helm does this internally but it seems to expect that he `charts` directory is removed before putting the just pulled versions in place. Not sure if it's related to this change, but going to test this now on a chart that uses the chartmuseum way of helm charts available to see if it also happens there.

**Comment by WyriHaximus** [comment]
> but going to test this now on a chart that uses the chartmuseum way of helm charts available to see if it also happens there.

@jdolitsky Ok just tested this and there is no issue with the helm I build from this PR using charts from a chartmuseum kind of repository

**Comment by cobexer** [comment]
There is an important loss of functionality here compared to old repositories:
`oci://localhost:3000/myregistry` may be a valid OCI registry in a CI/Test system, but production may want to store it's helm charts in something like: `oci://registry.prod.example.com/myregistry` - while our customer may want to store our Helm Charts in their own internal OCI registry...

This syntax needs a way of saying "in the same repository/registry as the parent chart" or "in any of the configured repositories/registries"

Am I missing something?

**Comment by jdolitsky** [comment]
@cobexer - you're asking for symbolic key (such as `@myrepo`), which is dynamic depending on environment?

In any case - I don't believe this PR does not prevent that feature being added in the future.

If you want, please make PR to modify this document: https://github.com/helm/community/blob/master/hips/hip-0006.md

**Comment by darren-bell-nanthealth** [comment]
Hi. Is there a potential release date for this feature? I'm hoping to use it for a DR scenario. Thanks

**Review by jdolitsky — APPROVED** [review]

**Review by mattfarina — CHANGES_REQUESTED** [review]
When I tried to run `helm dep up` (my local chart has a dependency on a chart in an OCI repo) more than once I got the following error if the chart already existed in the target location...

```
Error: unable to move local charts to charts dir: link error: cannot rename tmpcharts/test-chart to charts/test-chart: rename tmpcharts/test-chart charts/test-chart: file exists
```

Shouldn't it remove the old version and replace it with the new version?

I also noticed that `oci://` works even when the OCI experiment isn't enabled for dependencies in the `Chart.yaml` file. Is this intentional?

I've not completed my whole review but I wanted to pass on the couple issues and question I had right away.

File: pkg/getter/getter_test.go
Comment (pmengelbert) [reply pkg/getter/getter_test.go:60]: > I've not completed my whole review but I wanted to pass on the couple issues and question I had right away.

@mattfarina Thanks so much for the review. I just pushed a commit that fixes the bug you pointed out and gives the proper error message in `getter_test.go`. Let me know what you think of the changes, and I look forward to the rest of your review!

File: internal/experimental/registry/client.go
Comment (mattfarina) [inline internal/experimental/registry/client.go:149]: Note to reviewers... this function is internal and part of the experiment. It's ok to change it.

**Review by technosophos — APPROVED** [review]
This looks to me like a clear step forward on the experimental OCI front. I am good with merging/shipping this iteration.

**Comment by pmengelbert** [comment]
@mattfarina would you do us a favor and mark the requested changes as resolved? Once that's taken care of we can finally get this merged! Really happy to see this moving forward

**Comment by technosophos** [comment]
@pmengelbert we're actually both working on it right now

**Review by mattfarina — APPROVED** [review]

**Pull request merged**
